// ask-cli/test/unit/clients/smapi-client-test/resources/history.js
'use strict';

const expect = require('chai').expect;
const sinon = require('sinon');

const oauthWrapper = require('@src/utils/oauth-wrapper');
const CONSTANTS = require('@src/utils/constants');

const noop = () => {};

module.exports = (smapiClient) => {
    describe('# skill history APIs', () => {
        const TEST_SKILL_ID = 'skillId';
        const TEST_QUERRY_PARAMETERS = { foo: 'bar' };
    
        beforeEach(() => {
            sinon.stub(oauthWrapper, 'tokenRefreshAndRead');
        });
    
        [
            {
                testCase: 'get-utterance-data without query parameters',
                apiFunc: smapiClient.skill.history.getUtteranceData,
                parameters: [TEST_SKILL_ID, null, noop],
                expectedOptions: {
                    url: `${CONSTANTS.SMAPI.ENDPOINT}/v1/skills/${TEST_SKILL_ID}/history/intentRequests?`,
                    method: CONSTANTS.HTTP_REQUEST.VERB.GET,
                    headers: {},
                    body: null,
                    json: false
                }
            },
            {
                testCase: 'get-utterance-data',
                apiFunc: smapiClient.skill.history.getUtteranceData,
                parameters: [TEST_SKILL_ID, TEST_QUERRY_PARAMETERS, noop],
                expectedOptions: {
                    url: `${CONSTANTS.SMAPI.ENDPOINT}/v1/skills/${TEST_SKILL_ID}/history/intentRequests?foo=bar`,
                    method: CONSTANTS.HTTP_REQUEST.VERB.GET,
                    headers: {},
                    body: null,
                    json: false
                }
            }
        ].forEach(({ testCase, apiFunc, parameters, expectedOptions }) => {
            it (`| call ${testCase} successfully`, (done) => {
                // setup
                oauthWrapper.tokenRefreshAndRead.callsFake(noop);
                // call
                apiFunc(...parameters);
                // verify
                expect(oauthWrapper.tokenRefreshAndRead.called).equal(true);
                expect(oauthWrapper.tokenRefreshAndRead.args[0][0]).deep.equal(expectedOptions);
                done();
            });
        });
    
        afterEach(() => {
            oauthWrapper.tokenRefreshAndRead.restore();
        });
    });
};
