// ask-cli/test/functional/utils/constants.js
'use strict';
module.exports.apis = {
    GET_MODEL: 'get-model ',
    UPDATE_MODEL: 'update-model ',
    HEAD_MODEL: 'head-model ',
    CREATE_SKILL: 'create-skill ',
    GET_SKILL: 'get-skill ',
    UPDATE_SKILL: 'update-skill ',
    LIST_SKILL: 'list-skills ',
    GET_SKILL_STATUS: 'get-skill-status ',
    DELETE_SKILL: 'delete-skill ',
    GET_SKILL_ENABLEMENT: 'get-skill-enablement ',
    DISABLE_SKILL: 'disable-skill ',
    ENABLE_SKILL: 'enable-skill ',
    WITHDRAW: 'withdraw ',
    SUBMIT: 'submit ',
    DELETE_ACCOUNT_LINKING: 'delete-account-linking ',
    CREATE_ACCOUNT_LINKING: 'create-account-linking',
    GET_ACCOUNT_LINKING: 'get-account-linking ',
    LIST_VENDORS: 'list-vendors '

};

module.exports.hightLevel = {
    ASK_PREFIX: 'ask ',
    INIT: 'init ',
    NEW: 'new ',
    DEPLOY: 'deploy ',
    CLONE: 'clone ',
    SIMULATE: 'simulate ',
    VALIDATE: 'validate ',
    STATUS: 'status ',
    DIFF: 'diff ',
    LAMBDA: 'lambda ',
    API: 'api ',
    ADD: 'add ',
    REMOVE: 'remove ',
    UTIL: 'util '
}

module.exports.paramValue = {
    TEMPLATE: 'template',
    URL: 'url',
    SKILL_NAME: 'skill-name',
    PROFILE: 'profile',
    LAMNDA_NAME: 'lambda-name',
    SKILL_ID: 'skill id: ',
    TARGET: 'target'
}

module.exports.testData = {
    PROFILE: 'testProfile',
    DEVELOPMENT_STAGE: 'development',
    LIVE_STAGE: 'live',
    DEFAULT: 'default',
    INVALID_SKILL_ID: 'amzn1.ask.skill.e70e9c4a-8e5b-4789-a147-6487e100dfd',
    CLONE_SKILL_ID: 'amzn1.ask.skill.a089e56a-e453-40c9-9cb9-40b90e301c56',
    INVALID_PROFILE: 'invalidProfile',
    INVALID_STAGE: 'invalidStage',
    SKILL_NAME: 'cliIntegrationTest'
}

module.exports.locales = {
    EN_US: 'en-US'
}
